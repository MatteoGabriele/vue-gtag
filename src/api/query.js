import options from "@/options";

const query = (...args) => {
  if (typeof window === "undefined") {
    return;
  }

  window[options.globalObjectName](...args);
};

export default query;
