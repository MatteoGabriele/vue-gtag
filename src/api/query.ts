import { getSettings } from "@/settings";
import isServer from "@/utils";

export type QueryParams = Parameters<Gtag.Gtag>;

export default function query(...args: QueryParams) {
  const { dataLayerName, gtagName } = getSettings();

  if (isServer()) {
    return;
  }

  window[dataLayerName] = window[dataLayerName] || [];

  // biome-ignore lint/suspicious/noExplicitAny:
  const gtag = (...args: any[]) => window[dataLayerName].push(args);

  window[gtagName] = window[gtagName] || gtag;

  window[gtagName](...args);
}
