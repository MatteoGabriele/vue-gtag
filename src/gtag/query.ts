type QueryParams = Parameters<Gtag.Gtag>;

export default function query(...args: QueryParams): void {
  console.log(args)
}
