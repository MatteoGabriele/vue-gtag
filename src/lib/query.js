import { options } from "../install";

export default function(method, ...args) {
  if (typeof window === "undefined") {
    return;
  }

  if (window[options.globalObjectName]) {
    window[options.globalObjectName](method, ...args);
  } else {
    console.warn("gtag is not loaded yet")
  }

}