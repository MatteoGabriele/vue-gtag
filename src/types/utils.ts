export type Tail<T> = T extends [infer _, ...infer U] ? U : never;
