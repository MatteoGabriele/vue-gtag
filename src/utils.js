import { isPlainObject } from "@vue/shared";

export const load = (url, options = {}) => {
  return new Promise((resolve, reject) => {
    if (typeof document === "undefined") {
      return;
    }

    const head = document.head || document.getElementsByTagName("head")[0];
    const script = document.createElement("script");

    script.async = true;
    script.src = url;
    script.defer = options.defer;

    if (options.preconnectOrigin) {
      const link = document.createElement("link");

      link.href = options.preconnectOrigin;
      link.rel = "preconnect";

      head.appendChild(link);
    }

    head.appendChild(script);

    script.onload = resolve;
    script.onerror = reject;
  });
};

export const mergeDeep = (target, ...sources) => {
  if (!sources.length) {
    return target;
  }

  const source = sources.shift();

  if (!isPlainObject(target) || !isPlainObject(source)) {
    return;
  }

  for (const key in source) {
    if (isPlainObject(source[key])) {
      if (!target[key]) {
        Object.assign(target, { [key]: {} });
      }

      mergeDeep(target[key], source[key]);
    } else {
      Object.assign(target, { [key]: source[key] });
    }
  }

  return mergeDeep(target, ...sources);
};

export const isBrowser = () => {
  if (typeof window === "undefined" || typeof document === "undefined") {
    return false;
  }

  return true;
};

export const warn = (text, shouldLog = true) => {
  if (!isBrowser() || process.env.NODE_ENV === "production") {
    return;
  }

  if (!shouldLog) {
    return;
  }

  console.warn(`[vue-gtag] ${text}`);
};

export const validateScreenviewShape = (obj = {}) => {
  warn(
    `Missing "appName" property inside the plugin options.`,
    obj.app_name == null
  );

  warn(`Missing "name" property in the route.`, obj.screen_name == null);

  return obj;
};

export function getPathWithBase(path = "", base = "") {
  const pathAsArray = path.split("/");
  const baseAsArray = base.split("/");

  if (pathAsArray[0] === "" && base[base.length - 1] === "/") {
    pathAsArray.shift();
  }

  return baseAsArray.join("/") + pathAsArray.join("/");
}
