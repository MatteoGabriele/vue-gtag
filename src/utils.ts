export async function injectScript(
  url: string,
  options?: {
    preconnectOrigin?: string;
    defer?: boolean;
  },
): Promise<void> {
  return new Promise((resolve, reject) => {
    if (typeof document === "undefined") {
      return;
    }

    const head = document.head || document.getElementsByTagName("head")[0];
    const script = document.createElement("script");

    script.async = true;
    script.src = url;

    if (options?.defer) {
      script.defer = true;
    }

    if (options?.preconnectOrigin) {
      const link = document.createElement("link");

      link.href = options.preconnectOrigin;
      link.rel = "preconnect";

      head.appendChild(link);
    }

    head.appendChild(script);

    script.onload = () => resolve;
    script.onerror = reject;
  });
}
