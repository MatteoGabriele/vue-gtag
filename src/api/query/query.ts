import { getSettings } from "@/core/settings";
import type { Gtag } from "@/types/gtag";
import { isServer } from "@/utils";

export type QueryParams = Parameters<Gtag>;

declare global {
  interface Window {
    // biome-ignore lint/suspicious/noExplicitAny:
    [key: string]: any | any[];
  }
}

export function query(...args: QueryParams) {
  const { dataLayerName, gtagName } = getSettings();

  if (isServer()) {
    return;
  }

  window[dataLayerName] = window[dataLayerName] || [];

  // biome-ignore lint/complexity/useArrowFunction:
  window[gtagName] = function () {
    // biome-ignore lint/style/noArguments:
    window[dataLayerName].push(arguments);
  };

  window[gtagName](...args);
}
