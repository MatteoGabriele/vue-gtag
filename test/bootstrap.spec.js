import { install } from "./utils";
import * as utils from "@/utils";
import MockDate from "mockdate";
import query from "@/api/query";

jest.mock("@/api/query");

MockDate.set("06-03-1997 10:00:00");

describe("boostrap", () => {
  beforeEach(() => {
    jest.spyOn(utils, "load").mockResolvedValue();
  });

  test("load gtag script", () => {
    install({
      config: {
        id: "UA-123456-7",
      },
    });

    expect(utils.load).toHaveBeenCalledWith(
      "https://www.googletagmanager.com/gtag/js?id=UA-123456-7&l=dataLayer"
    );
  });

  test("define library and dataLayer", () => {
    install();

    expect(typeof window.gtag).toBe("function");
    expect(Array.isArray(window.dataLayer)).toBe(true);
  });

  test("define custom library and dataLayer", () => {
    install({
      globalDataLayerName: "myCustomDatalayer",
      globalObjectName: "myCustomGtag",
    });

    expect(typeof window.myCustomGtag).toBe("function");
    expect(Array.isArray(window.myCustomDatalayer)).toBe(true);
  });

  test("first dataLayer item is a date", () => {
    install();
    expect(window.dataLayer[0]).toMatchSnapshot();
  });

  test("gtag configuration hit", () => {
    install({
      config: {
        id: "UA-123456-7",
      },
    });

    expect(query).toHaveBeenCalledWith("config", "UA-123456-7", {
      send_page_view: false,
    });
  });
});
