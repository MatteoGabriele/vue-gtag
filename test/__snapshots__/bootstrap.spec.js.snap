// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`boostrap first dataLayer item is a date 1`] = `
Arguments [
  "js",
  1997-06-03T08:00:00.000Z,
]
`;
